### DISCUSSION TITLE
ModuleNotFoundError after freezing

### DISCUSSION THREAD

**Discussion by BaylorWong** [marcelotduarte/cx_Freeze#1470] [Q&A] [answered]
## Description
it works fine if i run NinjaTurtle.py by using IDE. 
But I try several ways to freeze with Cx_freeze it always end of ModuleNotFoundError or ImportError.I don't know what to do
as you can see,if i run NinjaTurtle.exe after freezing by this setup.py.The Error shows below.
![image](https://user-images.githubusercontent.com/72554492/163145264-0930fc9d-70a0-4793-a5c6-6fcbcf062438.png)
**it would be very helpful if you can give me some solution.Thanks!**
## setup.py
```
import os
import sys
from cx_Freeze import setup, Executable

current_path = os.path.dirname(__file__)
pkg_path = os.path.join(os.path.abspath(os.path.join(os.getcwd(), current_path, '..', '..')), 'pkg')
# ui_path = os.path.join(os.path.abspath(os.path.join(os.getcwd(), current_path, 'ui')))
sys.path.append(pkg_path)
# sys.path.append(ui_path)
print(sys.path)

base = None
if sys.platform == "win32":
    base = "Win32GUI"

options = {
    'build_exe': {
        'path': sys.path,
        'packages': ['service_controller',
                     'client_controller',
                     'config_operator'],
        # 'includes': ['client_controller.ClientController',
        #              'config_operator.ConfigOperator',
        #              'service_controller.ServiceControler'],
        # exclude packages that are not really needed
        "excludes": [],
        'include_files': ['Ninja-48.ico']
    }
}

setup(name="NinjaTurtle",
      version="0.1",
      description="Develop by Baylor",
      options=options,
      executables=[Executable("NinjaTurtle.py", icon='Ninja-48.ico', base=base)])
```
## NinjaTurtle's import
```
import os
import sys
import shutil
import socket
from functools import partial
from PySide6 import QtWidgets, QtCore
from PySide6.QtWidgets import QMessageBox, QLineEdit, QLabel, QPushButton, QComboBox, QCheckBox
from PySide6.QtGui import QIcon
from pkg.client_controller import ClientController
from pkg.config_operator import ConfigOperator
from pkg.service_controller import ServiceControler
import ninja
```
## code structure
![image](https://user-images.githubusercontent.com/72554492/163142529-46a430cc-8fae-4809-9f6e-5e4375d10516.png)

**Comment by marcelotduarte** [comment, marked as answer]
Usually, setup.py would be above the structure, but the location doesn't matter as long as the packages are installed.
So, if you are using pip, in the pkg directory you install it as a package (you have to create a basic setup.py from setuptools), you can install it with a commando like:
`pip install -e .`
It appears in the `pip list` as an installed package, you add it to cx_Freeze's setup.py in the package list (and you can remove the hack within sys.path).

**Reply by BaylorWong** [reply]
Thanks.The solution you recommend seems difficulte to me so I try to change my structure,and it works finally.

```
options = {
    'build_exe': {
        'path': sys.path,
        # 'packages': ['pkg'],
        'packages': ['pkg.service_controller',
                     'pkg.client_controller',
                     'pkg.config_operator',
                     'pkg.ui.ninja'],
        # 'includes': ['client_controller.ClientController',
        #              'config_operator.ConfigOperator',
        #              'service_controller.ServiceControler'],
        # exclude packages that are not really needed
        "excludes": ['email', 'html', 'http'],
        'include_files': ['Ninja-48.ico']
    }
}
```

![image](https://user-images.githubusercontent.com/72554492/163342678-41c0e8ad-099f-4df2-ad34-62687fa8eaf9.png)

But there is a extra directory `lib` has been created after freezing

![image](https://user-images.githubusercontent.com/72554492/163343707-3ee1e621-8cb3-44c5-8ba6-52e6a8c89b64.png)

**Reply by marcelotduarte** [reply]
This extra directory has files?

**Reply by BaylorWong** [reply]
![image](https://user-images.githubusercontent.com/72554492/163908031-6222cc2c-e2f1-41dd-8e2d-0cb9a27dc0e4.png)

**Reply by marcelotduarte** [reply]
I misunderstood the previous comment. A subdirectory 'lib' is the expected behavior.

**Reply by BaylorWong** [reply]
I mean there are two `lib` inside the `build` directory after freeze.But It should be one `lib` as expect
